internal/stream/notification_stream.go

```diff
@@ -0,0 +1,75 @@
+package stream
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/openshift/assisted-service/internal/common"
+	"github.com/sirupsen/logrus"
+)
+
+//go:generate mockgen -source=notification_stream.go -package=stream -destination=mock_notification_stream.go
+
+type Notifier interface {
+	Notify(ctx context.Context, notifiable common.Notifiable) error
+	Close()
+}
+
+type Envelope struct {
+	Name     string
+	Payload  interface{}
+	Metadata interface{}
+}
+
+type NotificationStream struct {
+	metadata interface{}
+	writer   StreamWriter
+	log      logrus.FieldLogger
+}
+
+func NewNotificationStream(writer StreamWriter, logger logrus.FieldLogger, metadata interface{}) *NotificationStream {
+	return &NotificationStream{
+		writer:   writer,
+		metadata: metadata,
+		log:      logger,
+	}
+
+}
+
+func (s *NotificationStream) Notify(ctx context.Context, notifiable common.Notifiable) error {
+	if s.writer == nil {
+		return nil
+	}
+	if notifiable == nil || reflect.ValueOf(notifiable).IsNil() {
+		return fmt.Errorf("trying to notify on nil notifiable")
+	}
+	key := ""
+	clusterID := notifiable.GetClusterID()
+	if clusterID != nil {
+		key = clusterID.String()
+	}
+
+	envelope := &Envelope{
+		Name:     notifiable.GetNotificationType(),
+		Payload:  notifiable.GetResource(),
```

Comment:
Is changing this to an empty interface really necessary? Does something not work if you just pass it as `common.Notifiable` type directly?

Reply:
I understand the concerns, however we need to pass the model here, not the wrapper, because:
* we need to leverage already-filtered fields
* we need to leverage well-defiined json conversion (named fields and such)
* but most importantly, we don't want to pass every column of the database, but rather the fields in the API. This would allow us to keep mindful retro compatibility (or break it, but still mindfully)

I would love to type check this, but I couldn't find a way to do it cleanly (that is not stretching it like adding a fake interface on some autogenerated methods, but I am not very comfortable with this as I am not sure what guarantees we have on such methods)